File: apollo-compiler/src/test/graphql/com/example/mutation_create_review_semantic_naming/type/ReviewInput.java
package com.example.mutation_create_review_semantic_naming.type;

import com.apollographql.apollo.api.InputFieldMarshaller;
import com.apollographql.apollo.api.InputFieldWriter;
import java.io.IOException;
import java.lang.IllegalStateException;
import java.lang.Integer;
import java.lang.Object;
import java.lang.Override;
import java.lang.String;
import java.util.List;
import javax.annotation.Generated;
import javax.annotation.Nonnull;
import javax.annotation.Nullable;

@Generated("Apollo GraphQL")
public final class ReviewInput {
  private final int stars;

  private final @Nullable Integer nullableIntFieldWithDefaultValue;

  private final @Nullable String commentary;

  private final @Nonnull ColorInput favoriteColor;

  private final @Nullable Episode enumWithDefaultValue;

  private final @Nullable Episode nullableEnum;

  private final @Nullable List<Object> listOfCustomScalar;

  private final @Nullable List<Episode> listOfEnums;

  ReviewInput(int stars, @Nullable Integer nullableIntFieldWithDefaultValue,
      @Nullable String commentary, @Nonnull ColorInput favoriteColor,
      @Nullable Episode enumWithDefaultValue, @Nullable Episode nullableEnum,
      @Nullable List<Object> listOfCustomScalar, @Nullable List<Episode> listOfEnums) {
    this.stars = stars;
    this.nullableIntFieldWithDefaultValue = nullableIntFieldWithDefaultValue;
    this.commentary = commentary;
    this.favoriteColor = favoriteColor;
    this.enumWithDefaultValue = enumWithDefaultValue;
    this.nullableEnum = nullableEnum;
    this.listOfCustomScalar = listOfCustomScalar;
    this.listOfEnums = listOfEnums;
  }

  /**
   * 0-5 stars
   */
  public int stars() {
    return this.stars;
  }

  /**
   * for test purpose only
   */
  public @Nullable Integer nullableIntFieldWithDefaultValue() {
    return this.nullableIntFieldWithDefaultValue;
  }

  /**
   * Comment about the movie, optional
   */
  public @Nullable String commentary() {
    return this.commentary;
  }

  /**
   * Favorite color, optional
   */
  public @Nonnull ColorInput favoriteColor() {
    return this.favoriteColor;
  }

  /**
   * for test purpose only
   */
  public @Nullable Episode enumWithDefaultValue() {
    return this.enumWithDefaultValue;
  }

  /**
   * for test purpose only
   */
  public @Nullable Episode nullableEnum() {
    return this.nullableEnum;
  }

  /**
   * for test purpose only
   */
  public @Nullable List<Object> listOfCustomScalar() {
    return this.listOfCustomScalar;
  }

  /**
   * for test purpose only
   */
  public @Nullable List<Episode> listOfEnums() {
    return this.listOfEnums;
  }

  public static Builder builder() {
    return new Builder();
  }

  public InputFieldMarshaller marshaller() {
    return new InputFieldMarshaller() {
      @Override
      public void marshal(InputFieldWriter writer) throws IOException {
        writer.writeInt("stars", stars);
        writer.writeInt("nullableIntFieldWithDefaultValue", nullableIntFieldWithDefaultValue);
        writer.writeString("commentary", commentary);
        writer.writeObject("favoriteColor", favoriteColor.marshaller());
        writer.writeString("enumWithDefaultValue", enumWithDefaultValue != null ? enumWithDefaultValue.name() : null);
        writer.writeString("nullableEnum", nullableEnum != null ? nullableEnum.name() : null);
        writer.writeList("listOfCustomScalar", listOfCustomScalar != null ? new InputFieldWriter.ListWriter() {
          @Override
          public void write(InputFieldWriter.ListItemWriter listItemWriter) throws IOException {
            for (Object $item : listOfCustomScalar) {
              listItemWriter.writeCustom(CustomType.DATE, $item);
            }
          }
        } : null);
        writer.writeList("listOfEnums", listOfEnums != null ? new InputFieldWriter.ListWriter() {
          @Override
          public void write(InputFieldWriter.ListItemWriter listItemWriter) throws IOException {
            for (Episode $item : listOfEnums) {
              listItemWriter.writeString($item.name());
            }
          }
        } : null);
      }
    };
  }

  public static final class Builder {
    private int stars;

    private @Nullable Integer nullableIntFieldWithDefaultValue = 10;

    private @Nullable String commentary;

    private @Nonnull ColorInput favoriteColor;

    private @Nullable Episode enumWithDefaultValue = Episode.JEDI;

    private @Nullable Episode nullableEnum;

    private @Nullable List<Object> listOfCustomScalar;

    private @Nullable List<Episode> listOfEnums;

    Builder() {
    }

    /**
     * 0-5 stars
     */
    public Builder stars(int stars) {
      this.stars = stars;
      return this;
    }

    /**
     * for test purpose only
     */
    public Builder nullableIntFieldWithDefaultValue(@Nullable Integer nullableIntFieldWithDefaultValue) {
      this.nullableIntFieldWithDefaultValue = nullableIntFieldWithDefaultValue;
      return this;
    }

    /**
     * Comment about the movie, optional
     */
    public Builder commentary(@Nullable String commentary) {
      this.commentary = commentary;
      return this;
    }

    /**
     * Favorite color, optional
     */
    public Builder favoriteColor(@Nonnull ColorInput favoriteColor) {
      this.favoriteColor = favoriteColor;
      return this;
    }

    /**
     * for test purpose only
     */
    public Builder enumWithDefaultValue(@Nullable Episode enumWithDefaultValue) {
      this.enumWithDefaultValue = enumWithDefaultValue;
      return this;
    }

    /**
     * for test purpose only
     */
    public Builder nullableEnum(@Nullable Episode nullableEnum) {
      this.nullableEnum = nullableEnum;
      return this;
    }

    /**
     * for test purpose only
     */
    public Builder listOfCustomScalar(@Nullable List<Object> listOfCustomScalar) {
      this.listOfCustomScalar = listOfCustomScalar;
      return this;
    }

    /**
     * for test purpose only
     */
    public Builder listOfEnums(@Nullable List<Episode> listOfEnums) {
      this.listOfEnums = listOfEnums;
      return this;
    }

    public ReviewInput build() {
      if (favoriteColor == null) throw new IllegalStateException("favoriteColor can't be null");
      return new ReviewInput(stars, nullableIntFieldWithDefaultValue, commentary, favoriteColor, enumWithDefaultValue, nullableEnum, listOfCustomScalar, listOfEnums);
    }
  }
}

File: apollo-compiler/src/test/graphql/com/example/input_object_type/type/ReviewInput.kt
package com.example.input_object_type.type

import com.apollographql.apollo.api.Input
import com.apollographql.apollo.api.InputFieldMarshaller
import com.apollographql.apollo.api.InputFieldWriter
import com.apollographql.apollo.api.InputType
import java.util.Date
import javax.annotation.Generated
import kotlin.Boolean
import kotlin.Int
import kotlin.String
import kotlin.Suppress
import kotlin.collections.List

/**
 * The input object sent when someone is creating a new review
 * @param stars 0-5 stars
 * @param nullableIntFieldWithDefaultValue for test purpose only
 * @param commentary Comment about the movie, optional
 * @param favoriteColor Favorite color, optional
 * @param enumWithDefaultValue for test purpose only
 * @param nullableEnum for test purpose only
 * @param listOfCustomScalar for test purpose only
 * @param customScalar for test purpose only
 * @param listOfEnums for test purpose only
 * @param listOfInt for test purpose only
 * @param listOfString for test purpose only
 * @param booleanWithDefaultValue for test purpose only
 * @param listOfListOfString for test purpose only
 * @param listOfListOfEnum for test purpose only
 * @param listOfListOfCustom for test purpose only
 * @param listOfListOfObject for test purpose only
 * @param capitalizedField for test purpose only
 */
@Generated("Apollo GraphQL")
@Suppress("NAME_SHADOWING", "LocalVariableName")
class ReviewInput(
    val stars: Int,
    val nullableIntFieldWithDefaultValue: Input<Int> = Input.optional(10),
    val commentary: Input<String> = Input.optional(null),
    val favoriteColor: ColorInput,
    val enumWithDefaultValue: Input<Episode> = Input.optional(Episode.safeValueOf("JEDI")),
    val nullableEnum: Input<Episode> = Input.optional(null),
    val listOfCustomScalar: Input<List<Date?>> = Input.optional(null),
    val customScalar: Input<Date> = Input.optional(null),
    val listOfEnums: Input<List<Episode?>> = Input.optional(null),
    val listOfInt: Input<List<Int?>> = Input.optional(listOf(1, 2, 3)),
    val listOfString: Input<List<String?>> = Input.optional(listOf("test1", "test2", "test3")),
    val booleanWithDefaultValue: Input<Boolean> = Input.optional(true),
    val listOfListOfString: Input<List<List<String?>?>> = Input.optional(null),
    val listOfListOfEnum: Input<List<List<Episode?>?>> = Input.optional(null),
    val listOfListOfCustom: Input<List<List<Date?>?>> = Input.optional(null),
    val listOfListOfObject: Input<List<List<ColorInput?>?>> = Input.optional(null),
    val capitalizedField: Input<String> = Input.optional(null)
) : InputType {
    override fun marshaller(): InputFieldMarshaller = InputFieldMarshaller { writer ->
        writer.writeInt("stars", stars)
        if (nullableIntFieldWithDefaultValue.defined) writer.writeInt("nullableIntFieldWithDefaultValue", nullableIntFieldWithDefaultValue.value)
        if (commentary.defined) writer.writeString("commentary", commentary.value)
        writer.writeObject("favoriteColor", favoriteColor.marshaller())
        if (enumWithDefaultValue.defined) writer.writeString("enumWithDefaultValue", enumWithDefaultValue.value?.rawValue)
        if (nullableEnum.defined) writer.writeString("nullableEnum", nullableEnum.value?.rawValue)
        if (listOfCustomScalar.defined) {
            writer.writeList("listOfCustomScalar", listOfCustomScalar.value?.let { value ->
                InputFieldWriter.ListWriter { listItemWriter ->
                    value.forEach { value ->
                        listItemWriter.writeCustom(CustomType.DATE, value)
                    }
                }
            })
        }
        if (customScalar.defined) writer.writeCustom("customScalar", CustomType.DATE, customScalar.value)
        if (listOfEnums.defined) {
            writer.writeList("listOfEnums", listOfEnums.value?.let { value ->
                InputFieldWriter.ListWriter { listItemWriter ->
                    value.forEach { value ->
                        listItemWriter.writeString(value?.rawValue)
                    }
                }
            })
        }
        if (listOfInt.defined) {
            writer.writeList("listOfInt", listOfInt.value?.let { value ->
                InputFieldWriter.ListWriter { listItemWriter ->
                    value.forEach { value ->
                        listItemWriter.writeInt(value)
                    }
                }
            })
        }
        if (listOfString.defined) {
            writer.writeList("listOfString", listOfString.value?.let { value ->
                InputFieldWriter.ListWriter { listItemWriter ->
                    value.forEach { value ->
                        listItemWriter.writeString(value)
                    }
                }
            })
        }
        if (booleanWithDefaultValue.defined) writer.writeBoolean("booleanWithDefaultValue", booleanWithDefaultValue.value)
        if (listOfListOfString.defined) {
            writer.writeList("listOfListOfString", listOfListOfString.value?.let { value ->
                InputFieldWriter.ListWriter { listItemWriter ->
                    value.forEach { value ->
                        listItemWriter.writeList{ listItemWriter-> 
                            value?.forEach { value ->
                                listItemWriter.writeString(value)
                            }
                        }
                    }
                }
            })
        }
        if (listOfListOfEnum.defined) {
            writer.writeList("listOfListOfEnum", listOfListOfEnum.value?.let { value ->
                InputFieldWriter.ListWriter { listItemWriter ->
                    value.forEach { value ->
                        listItemWriter.writeList{ listItemWriter-> 
                            value?.forEach { value ->
                                listItemWriter.writeString(value?.rawValue)
                            }
                        }
                    }
                }
            })
        }
        if (listOfListOfCustom.defined) {
            writer.writeList("listOfListOfCustom", listOfListOfCustom.value?.let { value ->
                InputFieldWriter.ListWriter { listItemWriter ->
                    value.forEach { value ->
                        listItemWriter.writeList{ listItemWriter-> 
                            value?.forEach { value ->
                                listItemWriter.writeCustom(CustomType.DATE, value)
                            }
                        }
                    }
                }
            })
        }
        if (listOfListOfObject.defined) {
            writer.writeList("listOfListOfObject", listOfListOfObject.value?.let { value ->
                InputFieldWriter.ListWriter { listItemWriter ->
                    value.forEach { value ->
                        listItemWriter.writeList{ listItemWriter-> 
                            value?.forEach { value ->
                                listItemWriter.writeObject(value?.marshaller())
                            }
                        }
                    }
                }
            })
        }
        if (capitalizedField.defined) writer.writeString("CapitalizedField", capitalizedField.value)
    }
}
